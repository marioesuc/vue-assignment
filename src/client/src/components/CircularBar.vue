<template>
  <div id="circularbar">
    <span id="currentValue">{{`${currentValue} ${unit}`}}</span>
    <VueSvgGauge
      :start-angle="-110"
      :end-angle="110"
      :value="currentValue"
      :separator-step="0"
      :min="0"
      :max="maxValue"
      :gauge-color="[{ offset: 0, color: '#347AB0'}, { offset: 100, color: '#8CDFAD'}]"
      :scale-interval="0.1"
      :transition-duration="0"
    />
  </div>
</template>

<script>
import { VueSvgGauge } from 'vue-svg-gauge'

export default {
  name: 'CircularBar',
  components: {
    VueSvgGauge,
  },
  props: {
    currentValue: Number,
    maxValue: Number,
    unit: String,
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
#circularbar {
  margin: 20px;
  width: 300px;
  height: 200px;
  text-align: center;
}
#currentValue {
  position: relative;
  top: 160px;
  color: green;
  font-size: 30px;
  font-weight: bold;
  text-align: center;
  margin: 0 auto;
}
</style>
