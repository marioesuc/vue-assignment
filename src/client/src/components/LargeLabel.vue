<template>
  <div id="largelabel">
    <span id="title">{{ `${title}` }}</span>
    <div id="labelValue">
      <span id="currentValue">{{ `${currentValue} ${unit}` }}</span>
    </div>
  </div>
</template>

<script>
export default {
  name: "LargeLabel",
  props: {
    title: String,
    currentValue: Number,
    unit: String
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import "@/common/Colors";

#largelabel {
  margin: 50px 20px;
}
#labelValue {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 230px;
  height: 60px;
  background-color: $largelabel-background;
  border-radius: 15px;
}
#title {
  display: flex;
  justify-content: center;
  color: $largelabel-title;
  font-size: 30px;
  font-weight: bold;
  margin: 0 0 12px;
}
#currentValue {
  color: $largelabel-value;
  font-size: 30px;
  font-weight: bold;
  text-align: center;
  margin: 0 auto;
}
</style>
