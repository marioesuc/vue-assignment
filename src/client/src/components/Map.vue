<template>
  <div id="map">
    <div id="mapShadow">
    </div>
    <l-map
      id="mapLayer"
      :zoom="zoom"
      :center="center"
      :options="mapOptions"
      style="height: 100%"
      @update:center="centerUpdate"
      @update:zoom="zoomUpdate"
    >
      <l-tile-layer :url="url" :attribution="attribution" />
      <l-marker :lat-lng="withPopup">
        <l-popup>
          <div @click="innerClick">
            I am a popup
            <p>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque
              sed pretium nisl, ut sagittis sapien. Sed vel sollicitudin nisi.
              Donec finibus semper metus id malesuada.
            </p>
          </div>
        </l-popup>
      </l-marker>
    </l-map>
  </div>
</template>

<script>
import { latLng } from "leaflet";
import { LMap, LTileLayer, LMarker, LPopup } from "vue2-leaflet";

export default {
  name: 'Map',
  components: {
    LMap,
    LTileLayer,
    LMarker,
    LPopup,
  },
  props: {
    latitude: Number,
    longitude: Number,
  },
  data() {
    return {
      zoom: 15,
      center: latLng(this.latitude, this.longitude),
      url: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
      attribution:
        '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
      withPopup: latLng(this.latitude, this.longitude),
      currentZoom: 11.5,
      currentCenter: latLng(this.latitude, this.longitude),
      mapOptions: {
        zoomSnap: 0.5
      },
    };
  },
  watch: {
    latitude() {
      this.center = latLng(this.latitude, this.longitude)
      this.withPopup = latLng(this.latitude, this.longitude)
      this.currentCenter = latLng(this.latitude, this.longitude)
    }
  },
  methods: {
    zoomUpdate(zoom) {
      this.currentZoom = zoom;
    },
    centerUpdate(center) {
      this.currentCenter = center;
    },
    showLongText() {
      this.showParagraph = !this.showParagraph;
    },
    innerClick() {
      alert('Click!');
    }
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
#map {
  position: relative;
  width: 800px;
  flex: 1;
  border: 3px solid white;
}
#mapLayer {
  border-radius: 15px;
  box-shadow: inset 0px 0px 10px 0px rgba(0,0,0,0.5);
}
#mapShadow {
  position: absolute;
  width: 100%;
  height: 100%;
  border-radius: 15px;
  box-shadow: inset 0px 0px 10px 0px rgba(0,0,0,0.5);
  z-index: 999;
}
</style>
