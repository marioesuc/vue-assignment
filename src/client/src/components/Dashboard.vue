<template>
  <div id="dashboard">
    <div class="row">
      <Map
        :latitude="data.gps.latitude"
        :longitude="data.gps.longitude"
      />
      <div class="column">
        <div class="row">
          <div class="column">
          <CircularBar
            :currentValue="data.speed"
            :maxValue="maxSpeed"
            :unit="'km/h'"
          />
          </div>
          <div class="column">
          <CircularBar
            :currentValue="data.soc"
            :maxValue="maxCharge"
            :unit="'%'"
          />
          </div>
        </div>
        <div class="row">
          <div class="column">
          <LargeLabel
            :currentValue="data.energy"
            :unit="'kW'"
          />
          </div>
          <div class="column">
            <LargeLabel
              :currentValue="data.odo"
              :unit="'km'"
            />
          </div>
        </div>
      </div>
    </div>
    <div class="row">
      <LinearChart
        :speed="data.speed"
        :soc="data.soc"
        :time="data.time"
      />
    </div>
  </div>
</template>

<script>
import Map from "./Map";
import CircularBar from "./CircularBar";
import LinearChart from "./LinearChart";
import LargeLabel from "./LargeLabel";
import { MAX_SPEED, MAX_CHARGE } from "@/common/Constants";

export default {
  name: "Dashboard",
  components: {
    Map,
    CircularBar,
    LinearChart,
    LargeLabel,
  },
  props: {
    data: Object
  },
  data() {
    return {
      maxSpeed: MAX_SPEED,
      maxCharge: MAX_CHARGE
    }
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
#dashboard {
  width: 1350px;
  background-color: white;
  margin: 30px auto;
  padding: 20px;
  box-shadow: 0px 5px 15px 0px rgba(0,0,0,0.75);
}
.row {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: 100%;
  justify-content: center;
}
.column {
  display: flex;
  flex-direction: column;
  flex: 1;
  justify-content: center;
  align-items: center;
}
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
</style>
