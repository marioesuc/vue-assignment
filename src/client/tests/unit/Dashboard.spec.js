import { shallowMount } from "@vue/test-utils";
import Dashboard from "@/components/Dashboard.vue";
import Map from "@/components/Map.vue";
import CircularBar from "@/components/CircularBar.vue";
import LargeLabel from "@/components/LargeLabel.vue";

describe("Dashboard.vue", () => {
  it("renders correctly", () => {
    // Stub dynamically loaded components replacing them with their real components
    const stubs = {
      Map,
      CircularBar,
      LargeLabel,
      // Stub LinearChart to fix failing tests because of canvas
      LinearChart: "<span></span>"
    };

    const propsData = {
      data: {
        time: 1511436338000,
        energy: 53.2,
        gps: {
          latitude: 52.093448638916016,
          longitude: 5.117378234863281
        },
        odo: 88526.413,
        speed: 10,
        soc: 72.8
      }
    };

    const wrapper = shallowMount(Dashboard, { propsData, stubs });
    expect(wrapper.find(Map).exists()).toBeTruthy();
    expect(wrapper.find(CircularBar).exists()).toBeTruthy();
    expect(wrapper.find(LargeLabel).exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
