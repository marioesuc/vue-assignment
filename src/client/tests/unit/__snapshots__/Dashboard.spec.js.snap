// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue renders correctly 1`] = `
<div id="dashboard">
  <div class="row">
    <div class="column">
      <div id="map">
        <l-map-stub options="[object Object]" center="LatLng(52.093449, 5.117378)" zoom="15" crs="[object Object]" id="mapLayer">
          <div id="mapShadow"></div>
          <l-tile-layer-stub pane="tilePane" attribution="&amp;copy; <a href=&quot;http://osm.org/copyright&quot;>OpenStreetMap</a> contributors" visible="true" opacity="1" zindex="1" tilesize="256" subdomains="abc" options="[object Object]" url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" tilelayerclass="function tileLayer(url, options) {
	return new TileLayer(url, options);
}"></l-tile-layer-stub>
          <l-marker-stub pane="markerPane" visible="true" options="[object Object]" latlng="LatLng(52.093449, 5.117378)" icon="[object Object]"></l-marker-stub>
        </l-map-stub>
      </div>
    </div>
    <div class="column">
      <div class="row">
        <div class="column">
          <div id="circularbar"><span id="currentValue" class="thresholdNotExceeded">10 km/h</span> <span id="title">speed</span>
            <vuesvggauge-stub value="10" min="0" max="100" startangle="-110" endangle="110" innerradius="60" separatorstep="0" separatorthickness="4" gaugecolor="[object Object],[object Object]" basecolor="#DDDDDD" easing="Circular.Out" scaleinterval="0.1" transitionduration="0"></vuesvggauge-stub>
          </div>
        </div>
        <div class="column">
          <div id="circularbar"><span id="currentValue" class="thresholdNotExceeded">72.8 %</span> <span id="title">state of charge</span>
            <vuesvggauge-stub value="72.8" min="0" max="100" startangle="-110" endangle="110" innerradius="60" separatorstep="0" separatorthickness="4" gaugecolor="[object Object],[object Object]" basecolor="#DDDDDD" easing="Circular.Out" scaleinterval="0.1" transitionduration="0"></vuesvggauge-stub>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="column">
          <div id="largelabel"><span id="title">energy</span>
            <div id="labelValue"><span id="currentValue">53.2 kW</span></div>
          </div>
        </div>
        <div class="column">
          <div id="largelabel"><span id="title">odometer</span>
            <div id="labelValue"><span id="currentValue">88526.413 km</span></div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="row"><span data="[object Object]" options="[object Object]" intervalenabled="" refreshrate="250"></span></div>
</div>
`;
